prusti-interface/src/environment/mod.rs

```diff
@@ -320,21 +337,65 @@ impl<'tcx> Environment<'tcx> {
             .is_some()
     }
 
-    /// Returns the `DefId` of the corresponding trait method
-    pub fn find_trait_method(&self, impl_def_id: ProcedureDefId) -> Option<DefId> {
+    /// Returns the `DefId` of the corresponding trait method, if any.
+    /// This should not be used to resolve calls (where substs are known): use
+    /// `find_trait_method_substs` instead!
+    pub fn find_trait_method(
+        &self,
+        impl_def_id: ProcedureDefId, // what are we calling?
+    ) -> Option<DefId> {
         self.tcx
             .impl_of_method(impl_def_id)
             .and_then(|impl_id| self.tcx.trait_id_of_impl(impl_id))
             .and_then(|trait_id| self.get_assoc_item(trait_id, self.tcx().item_name(impl_def_id)))
             .map(|assoc_item| assoc_item.def_id)
     }
 
+    /// If the given `impl_method_def_id` is an implementation of a trait
+    /// method, return the `DefId` of that trait method as well as an adapted
+    /// version of the callsite `impl_method_substs` substitutions.
+    pub fn find_trait_method_substs(
+        &self,
+        impl_method_def_id: ProcedureDefId, // what are we calling?
+        impl_method_substs: SubstsRef<'tcx>, // what are the substs on the call?
+    ) -> Option<(ProcedureDefId, SubstsRef<'tcx>)> {
+        let impl_def_id = self.tcx.impl_of_method(impl_method_def_id)?;
+        let trait_def_id = self.tcx.trait_id_of_impl(impl_def_id)?;
+
+        // the input method was in an impl for a trait: finding a `None` in
+        // further calls would be unexpected, so we unwrap instead of ?
+
+        let trait_method_def_id = self.get_assoc_item(
+            trait_def_id,
+            self.tcx().item_name(impl_method_def_id),
+        ).unwrap().def_id;
```

Comment:
I think it would be better to change the return type to `Result` and return `Result::Err` here to be able to show a user-friendly internal error message to the user instead of crashing.

Reply:
I don't want to clutter the interface. If we get an unwrap error now (after the two successful `?` unwraps just above), there is probably something wrong inside the compiler. The two calls establish that:

- the given method belongs to an `impl` block
- that `impl` block is implementing a trait

So we should definitely be able to get the associated item (or else what is the method implementing?) and we should definitely be able to get the trait ref (or else how could the `impl` block be implementing a trait?).

Reply:
In that case, I would clarify the comment to explicitly say that we expect a compiler bug in this case.